test/dns.go

```diff
@@ -0,0 +1,198 @@
+		if !ok {
+			// ^ 				start of line
+			// localhost\.		match literally
+			// ()* 				match between 0 and unlimited times
+			// [[:alnum:]]+\.	match single character in [a-zA-Z0-9] minimum one time and ending in . literally
+			reg := regexp.MustCompile(`^localhost\.([[:alnum:]]+\.)*`)
+			if matched := reg.MatchString(domain); !matched {
+				panic(fmt.Sprintf("domain not mocked: %s", domain))
+			}
+
+			addresses = []string{"127.0.0.1"}
+		}
+
+		for _, addr := range addresses {
+			msg.Answer = append(msg.Answer, &dns.A{
+				Hdr: dns.RR_Header{Name: domain, Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: 60},
+				A:   net.ParseIP(addr),
+			})
+		}
+	}
+	w.WriteMsg(&msg)
+}
+
+type DnsMockHandle struct {
+	id              string
+	mockServer      *dns.Server
+	ShutdownDnsMock func() error
+}
+
+func (h *DnsMockHandle) PushDomains(domainsMap map[string][]string, domainsErrorMap map[string]int) func() {
+	handler := h.mockServer.Handler.(*dnsMockHandler)
+	handler.muDomainsToAddresses.Lock()
+	defer handler.muDomainsToAddresses.Unlock()
+
+	dta := handler.domainsToAddresses
+	dte := handler.domainsToErrors
+
+	prevDta := map[string][]string{}
+	prevDte := map[string]int{}
+
+	for key, value := range dta {
+		prevDta[key] = value
+	}
+
+	for key, value := range dte {
+		prevDte[key] = value
+	}
+
+	pullDomainsFunc := func() {
+		handler := h.mockServer.Handler.(*dnsMockHandler)
+		handler.muDomainsToAddresses.Lock()
+		defer handler.muDomainsToAddresses.Unlock()
+
+		handler.domainsToAddresses = prevDta
+		handler.domainsToErrors = prevDte
+	}
+
+	for key, ips := range domainsMap {
+		addr, ok := dta[key]
+		if !ok {
+			dta[key] = ips
+		} else {
+			dta[key] = append(addr, ips...)
+		}
+	}
+
+	for key, rCode := range domainsErrorMap {
+		dte[key] = rCode
+	}
+
+	return pullDomainsFunc
+}
+
+// InitDNSMock initializes dns server on udp:0 address and replaces net.DefaultResolver in order
+// to route all dns queries within tests to this server.
+// InitDNSMock returns handle, which can be used to add/remove dns query mock responses or initialization error.
+func InitDNSMock(domainsMap map[string][]string, domainsErrorMap map[string]int) (*DnsMockHandle, error) {
+	addr, _ := net.ResolveUDPAddr("udp", ":0")
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return &DnsMockHandle{}, err
+	}
+
+	startResultChannel := make(chan error)
+	started := func() {
+		startResultChannel <- nil
+	}
+
+	mockServer := &dns.Server{PacketConn: conn, NotifyStartedFunc: started}
+	handle := &DnsMockHandle{id: time.Now().String(), mockServer: mockServer}
+
+	dnsMux := &dnsMockHandler{muDomainsToAddresses: sync.RWMutex{}}
+
+	if domainsMap != nil {
+		dnsMux.domainsToAddresses = domainsMap
+	} else {
+		dnsMux.domainsToAddresses = DomainsToAddresses
+	}
+
+	if domainsErrorMap != nil {
+		dnsMux.domainsToErrors = domainsErrorMap
+	}
+
+	mockServer.Handler = dnsMux
+
+	go func() {
+		err := mockServer.ActivateAndServe()
```

Comment:
might simplify `startResultChannel <- mockServer.ActivateAndServe()`

Reply:
fixed in [7f332e6](https://github.com/TykTechnologies/tyk/commit/7f332e6a861e73fc8eb846e0e9aadd5f96fc724e)